Non-Escapable Types (#2304)
swiftlang/swift-evolution: proposals/NNNN-non-escapable.md

}
f(NotEscapable()) // Ok to call with non-escapable argument
f(7) // Ok to call with escapable argument
```

This also permits the definition of types whose escapability varies depending on their generic arguments.
As with other conditional behaviors, this is expressed by using an extension to conditionally add a new capability to the type:

```
// Example: Conditionally Escapable generic type
// By default, Box is itself non-escapable
struct Box<T: ~Escapable>: ~Escapable {
  var t: T
}

// Box gains the ability to escape whenever its
// generic argument is Escapable
extension Box: Escapable when T: Escapable { }
```

**Note:**  There is no relationship between `Copyable` and `Escapable`.
Copyable or non-copyable types can be escapable or non-escapable.

#### Constraints on non-escapable local variables

A non-escapable value can be freely copied and passed into other functions as long as the usage can guarantee that the value does not persist beyond the current scope:

```
// Example: Local variable with non-escapable type
func borrowingFunc(_: borrowing NotEscapable) { ... }
func consumingFunc(_: consuming NotEscapable) { ... }

> func inoutFunc(_: inout NotEscapable) { ... }

Reviewer: It'd be good to add a case for async functions and async let here, it's barely mentioned in the proposal.

Author: I'll see what I can do about including async examples.

Until then, note that the comments about functions apply to _all_ functions, regardless of whether they are async or not, throwing or not.

Similarly, `async let` guarantees that the called function will return before the current function completes, so it's safe to pass a `~Escapable` value to a  function called with `async let`.

# Nonescapable Types

* Proposal: [SE-NNNN](NNNN-filename.md)
* Authors: [Andrew Trick](https://github.com/atrick), [Tim Kientzle](https://github.com/tbkka)
* Review Manager: TBD
* Status: **Awaiting implementation**
* Roadmap: [BufferView Language Requirements](https://forums.swift.org/t/roadmap-language-support-for-bufferview)
* Implementation: **Pending**
* Upcoming Feature Flag: `NonescapableTypes`
* Review: ([pitch](https://forums.swift.org/t/pitch-non-escapable-types-and-lifetime-dependency/69865))

## Introduction

We propose adding a new type constraint `~Escapable` for types that can be locally copied but cannot be assigned or transferred outside of the immediate context.
This complements the `~Copyable` types added with SE-0390 by introducing another set of compile-time-enforced lifetime controls that can be used for safe, highly-performant APIs.

In addition, these types will support lifetime-dependency constraints (being tracked in a separate proposal), that allow them to safely hold pointers referring to data stored in other types.

> This feature is a key requirement for the proposed `StorageView` type.

Reviewer 2: ```suggestion
This feature is a key requirement for the proposed `Span` family of types.
```

#### Closures and nonescapable values

Escaping closures cannot capture nonescapable values.
Nonescaping closures can capture nonescapable values subject only to the usual exclusivity restrictions.

Returning a nonescapable value from a closure requires explicit lifetime dependency annotations, as covered in the companion proposal.

#### Nonescapable values and concurrency

All of the requirements on use of nonescapable values as function arguments and return values also apply to async functions, including those invoked via `async let`.

The closures used in `Task.init`, `Task.detached`, or `TaskGroup.addTask` are escaping closures and therefore cannot capture nonescapable values.

#### Conditionally `Escapable` types

You can define types whose escapability varies depending on their generic arguments.
As with other conditional behaviors, this is expressed by using an extension to conditionally add a new capability to the type:

```swift
// Example: Conditionally Escapable generic type
// By default, Box is itself nonescapable
struct Box<T: ~Escapable>: ~Escapable {
  var t: T
}

// Box gains the ability to escape whenever its
// generic argument is Escapable

> extension Box: Escapable when T: Escapable { }

Reviewer 3: ```suggestion
extension Box: Escapable where T: Escapable { }
```

Author: Done.

```

This can be used in conjunction with other suppressible protocols.
For example, many general library types will need to be copyable and/or escapable following their contents.
Here's a compact way to declare such a type:
```swift
struct Wrapper<T: ~Copyable & ~Escapable> { ... }
extension Wrapper: Copyable where T: ~Escapable {}
extension Wrapper: Escapable where T: ~Copyable {}
```

The above declarations all in a single source file will result in a type `Wrapper` that is `Escapable` exactly when `T` is `Escapable` and `Copyable` exactly when `T` is `Copyable`.
To see why, first note that the explicit `extension Wrapper: Escapable` in the same source file implies that the original `struct Wrapper` must be `~Escapable` and similarly for `Copyable`, exactly as if the first line had been
```swift
struct Wrapper<T: ~Copyable & ~Escapable>: ~Copyable & ~Escapable {}
```

Now recall from SE-427 that suppressible protocols must be explicitly suppressed on type parameters in extensions.
This means that
```swift

> extension Wrapper: Copyable where T: ~Escapable {}

Reviewer 2: Per SE-0427, this should be an error; the conditional conformance needs to explicitly declare whether it wants T to be copiable. (I.e., the only legal way to spell the conditional conformance is to use the long form below.)

```
is exactly the same as
```swift
extension Wrapper: Copyable where T: Copyable & ~Escapable {}
```
which implies that `Wrapper` becomes `Copyable` when `T` is `Copyable`.
Finally, remember that `~Escapable` means that `Escapable` is not required, so
this condition on `Copyable` applies regardless of whether `T` is `Escapable` or not.

Similarly,
```swift

> extension Wrapper: Escapable where T: ~Copyable {}

Reviewer 2: Same here; SE-0427 implies that this should be an error.

Adding `~Escapable` to an existing concrete type is generally source-breaking because existing source code may rely on being able to escape values of this type.
Removing `~Escapable` from an existing concrete type is not generally source-breaking since it effectively adds a new capability, similar to adding a new protocol conformance.

## ABI compatibility

As above, existing code is unaffected by this change.
Adding or removing a `~Escapable` constraint on an existing type is an ABI-breaking change.

## Implications on adoption

Manglings and interface files will only record the lack of escapability.
This means that existing interfaces consumed by a newer compiler will treat all types as escapable.
Similarly, an old compiler reading a new interface will have no problems as long as the new interface does not contain any `~Escapable` types.

These same considerations ensure that escapable types can be shared between previously-compiled code and newly-compiled code.

Retrofitting existing generic types so they can support both escapable and nonescapable type arguments is possible with care.

## Future directions

> #### `StorageView` type

Reviewer 2: ```suggestion
#### `Span` family of types
```

> This proposal is being driven in large part by the needs of the `StorageView` type that has been discussed elsewhere.

Reviewer 2: ```suggestion
This proposal is being driven in large part by the needs of the `Span` type that has been discussed elsewhere.
```

Briefly, this type would provide an efficient universal “view” of array-like data stored in contiguous memory.
Since values of this type do not own any data but only refer to data stored elsewhere, their lifetime must be limited to not exceed that of the owning storage.
We expect to publish a sample implementation and proposal for that type very soon.

#### Lifetime dependency annotations

Nonescapable types have a set of inherent restrictions on how they can be passed as arguments, stored in variables, or returned from functions.
A companion proposal builds on this by supporting more detailed annotations that link the lifetimes of different objects.
This would allow, for example, a container to vend an iterator value that held a direct unmanaged pointer to the container's contents.
The lifetime dependency would ensure that such an iterator could not outlive the container to whose contents it referred.

```swift
// Example: Nonescaping iterator
struct NEIterator {
  // `borrow(container)` indicates that the constructed value
  // cannot outlive the `container` argument.

>   init(over container: MyContainer) -> borrow(container) Self {

Reviewer 3: Suggest updating syntax to match companion proposal :)

Author: Done.  (Thanks for the feedback!)

Apart from that, we expect almost all types to continue to be escapable in the future, so the negative marker reduces the overall burden.
It is also consistent with progressive disclosure:
Most new Swift programmers should not need to know details of how escapable types work, since that is the common behavior of most data types in most programming languages.
When developers use existing nonescapable types, specific compiler error messages should guide them to correct usage without needing to have a detailed understanding of the underlying concepts.
With our current proposal, the only developers who will need detailed understanding of these concepts are library authors who want to publish nonescapable types.

#### `Nonescapable` as a marker protocol

We considered introducing `Nonescapable` as a marker protocol indicating that the values of this type required additional compiler checks.
With that approach, you would define a conditionally-escapable type such as `Box` above in this fashion:

```swift
// Box does not normally require additional escapability checks
struct Box<T> {
  var t: T
}

// But if T requires additional checks, so does Box

> extension Box: Nonescapable when T: Nonescapable { }

Reviewer 3: ```suggestion
extension Box: Nonescapable where T: Nonescapable { }
```

Author: Done.
